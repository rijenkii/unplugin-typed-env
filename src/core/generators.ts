import type { Field } from "./field.js";

export function js(schema: Record<string, Field>): string {
  let jsSrc = "";
  for (const [fieldName, field] of Object.entries(schema)) {
    try {
      const fieldValue = field.parse(process.env[fieldName]);
      jsSrc += `export const ${fieldName} = ${fieldValue};\n`;
    } catch (cause) {
      throw new Error(
        `Error while parsing field ${JSON.stringify(fieldName)}`,
        { cause },
      );
    }
  }
  return jsSrc;
}

export function dts(schema: Record<string, Field>): string {
  let dtsSrc = "";
  dtsSrc += "/* eslint-disable */\n";
  dtsSrc += "/* prettier-ignore */\n";
  dtsSrc += "// @ts-nocheck\n";
  dtsSrc +=
    "// Generated by unplugin-typed-env. !! DO NOT MODIFY THIS FILE !!\n";
  dtsSrc += "// It's recommended to commit this file.\n";
  dtsSrc +=
    '// Make sure to add this file to your tsconfig.json file as an "includes" or "files" entry."\n';
  dtsSrc += "\n";
  dtsSrc += 'declare module "virtual:typed-env" {\n';

  for (const [fieldName, field] of Object.entries(schema)) {
    const fieldType = field.type;
    dtsSrc += `export const ${fieldName}: ${fieldType};\n`;
  }

  dtsSrc += "}\n";

  return dtsSrc;
}
