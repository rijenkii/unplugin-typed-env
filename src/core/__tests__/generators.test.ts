import { afterEach, describe, expect, it, vi } from "vitest";

import { dts, js } from "../generators.js";

afterEach(() => vi.unstubAllEnvs());

describe("dts", () => {
  it("generates empty module when schema is empty", () => {
    const result = dts({});
    expect(result).toMatchSnapshot();
  });

  it("generates module with fields when schema is not empty", () => {
    const result = dts({
      FIELD_NAME_1: { type: "fieldType | fieldOtherType", parse: () => "" },
      FIELD_NAME_2: { type: "string", parse: () => "" },
    });
    expect(result).toMatchSnapshot();
  });
});

describe("js", () => {
  it("generates empty module when schema is empty", () => {
    const result = js({});
    expect(result).toMatchSnapshot();
  });

  it("throws when field validation fails", () => {
    const parse = vi.fn(() => {
      throw new Error("Required");
    });
    vi.stubEnv("FIELD_NAME_1", "value");

    expect(() =>
      js({
        FIELD_NAME_1: { type: "", parse },
      }),
    ).toThrow(
      new Error('Error while parsing field "FIELD_NAME_1"', {
        cause: new Error("Required"),
      }),
    );
    expect(parse).toHaveBeenCalledExactlyOnceWith("value");
  });

  it("generates module with parsed fields when schema is not empty", () => {
    const parse1 = vi.fn(() => "parsedValue1");
    const parse2 = vi.fn(() => "parsedValue2");
    vi.stubEnv("FIELD_NAME_1", "value1");
    vi.stubEnv("FIELD_NAME_2", "value2");
    const result = js({
      FIELD_NAME_1: { type: "", parse: parse1 },
      FIELD_NAME_2: { type: "", parse: parse2 },
    });
    expect(parse1).toHaveBeenCalledExactlyOnceWith("value1");
    expect(parse2).toHaveBeenCalledExactlyOnceWith("value2");
    expect(result).toMatchSnapshot();
  });
});
