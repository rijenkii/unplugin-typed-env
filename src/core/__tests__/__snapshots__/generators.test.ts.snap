// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dts > generates empty module when schema is empty 1`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// Generated by unplugin-typed-env. !! DO NOT MODIFY THIS FILE !!
// It's recommended to commit this file.
// Make sure to add this file to your tsconfig.json file as an "includes" or "files" entry."

declare module "virtual:typed-env" {
}
"
`;

exports[`dts > generates module with fields when schema is not empty 1`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// Generated by unplugin-typed-env. !! DO NOT MODIFY THIS FILE !!
// It's recommended to commit this file.
// Make sure to add this file to your tsconfig.json file as an "includes" or "files" entry."

declare module "virtual:typed-env" {
export const FIELD_NAME_1: fieldType | fieldOtherType;
export const FIELD_NAME_2: string;
}
"
`;

exports[`js > generates empty module when schema is empty 1`] = `""`;

exports[`js > generates module with parsed fields when schema is not empty 1`] = `
"export const FIELD_NAME_1 = parsedValue1;
export const FIELD_NAME_2 = parsedValue2;
"
`;
